### ISSUE TITLE
lib-protobuf-3.0.0-host

### ISSUE THREAD

**Issue by Mohamad-Jaallouk** [ctuning/ck-env#117]
By Running for the first time: `ck run cmdgen:benchmark.object-detection.tflite-loadgen --verbose \
--library=tflite-edgetpu --model:=v1:v2 \
--scenario=singlestream --mode=performance --target_latency=20 \
--sut=rpi4coral`

protobuf-3.0.0/tests.sh
protobuf-3.0.0/update_file_lists.sh
protobuf-3.0.0/util/
protobuf-3.0.0/util/python/
protobuf-3.0.0/util/python/BUILD

[Patcher] Patching of the original code disabled

Cleaning ...

CMake configure flags:

-Dprotobuf_BUILD_TESTS=OFF -DCMAKE_INSTALL_LIBDIR=lib

Executing cmake ...
/home/pi/CK/local/script/download-and-install-package/install.sh: line 490: /home/pi/CK-TOOLS/tool-cmake-3.14.3-linux-32/cmake-3.14.3-Linux-x86_64/bin/cmake: cannot execute binary file: Exec format error
Error: configuring failed!

   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   CK detected a PROBLEM in the third-party CK package:

   CK package:           lib-protobuf-3.0.0-host
   Native package URL:   https://github.com/google/protobuf/archive
   CK repo:              ck-env

   CK repo URL:          https://github.com/ctuning/ck-env
   CK package URL:       https://github.com/ctuning/ck-env/tree/master/package/lib-protobuf-3.0.0-host

   CodeReef stable package URL: https://cKnowledge.io/c/package/lib-protobuf-3.0.0-host
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   Please, submit the *full* log to the authors of this external CK package
   at https://github.com/ctuning/ck-env/issues to collaboratively fix this problem!

CK error: [db25414b48b4ffb3] package installation failed!
==========================================================================================

**Comment by gfursin** [comment]
Hi @quake2005,

Thank you for your report!

The problem is that you installed cmake from the CK package for x86 while your host is Arm-based (RPi).

You can fix that as follows (I just tested that on my RPi4):
Delete your current CK env with wrong cmake:
```
ck rm env:* --tags=cmake
```
Build cmake using this CK package "tool-cmake":
```
ck install package:tool-cmake
```
It may take ~5..7 minutes and then you can resume CK MLPerf workflows ...

Please tell me if it works!

By the way, note that I've moved most of the CK ML developments to the [ck-ml](https://github.com/ctuning/ck-ml) repository - it aggregated all CK components from the community for AI/ML in one place to make it easier to test and use...

Thanks!

**Comment by gfursin** [comment]
Extra info: it takes ~35 minutes to build cmake ;) ...
Alternative is to install cmake with apt and then detect it with CK:

```
sudo apt install cmake
ck detect soft --tags=tool,cmake
```
I use Ubuntu 20.04 server 64-bit on my RPi4 and it installs 3.16.3 version which should be fine for MLPerf benchmarking ...

**Comment by quake2005** [comment]
I tried the second method and it worked!
Now I get another problem:

  -----------------------------------
  Resolving software dependencies ...

*** Dependency 1 = compiler (C++ compiler):

 ********
 WARNING: no registered CK environment was found for "C++ compiler" dependency with tags="compiler,lang-cpp,gcc,v9" and setup={"host_os_uoa": "linux-64", "target_os_uoa": "linux-64", "target_os_bits": "64"}

  Trying to automatically detect required software ...
    No software auto-detection scripts found for this software in CK :( ...

  Searching and installing CK software packages ...
    * tags:        compiler,lang-cpp,gcc,v9
    * or tags:     
    * no tags:     

    CK packages are not found for this software :( !


CK error: [db25414b48b4ffb3] no registered CK environment was found for "C++ compiler" dependency with 

 tags="compiler,lang-cpp,gcc,v9" and setup={"host_os_uoa": "linux-64", "target_os_uoa": "linux-64", "target_os_bits": "64"}!

I use Raspberry Pi OS 64-bit.
Thanks for your time!

**Comment by gfursin** [comment]
No worries! Can you please give me the link to the original instructions you are following, please? 

Normally, you should do the following before running the cmdgen:
```
ck detect soft:compiler.gcc
```
to detect GCC v9 on RPi4 and register in the CK env.

I will check it further ...

**Comment by quake2005** [comment]
My goal is to do an object detection benchmark on Pi4 with and without Coral USB on both 32-bit and 64-bit.
I tried:
```
ck detect soft --tags=compiler,gcc
ck detect soft --tags=compiler,llvm
```
Now I have many versions of GCC and installed and the environment is a bit messy. I will try again with Ubuntu 20.04 server 64-bit.
I follow: 
https://github.com/ctuning/ck-mlperf/blob/master/program/object-detection-tflite-loadgen/README.md
https://github.com/ctuning/ck-mlperf/blob/master/program/object-detection-tflite-loadgen/README.singlestream.md

Thanks!

**Comment by gfursin** [comment]
Sure! Thanks for the info! Let me check it.
Repo ctuning/ck-mlperf can be a bit outdated. I use ctuning/ck-ml ...

**Comment by gfursin** [comment]
Hi again, @quake2005 .

I tested EdgeTPU CK workflow on RPi4 with Ubuntu 20.04 64-bit server from scratch and it worked.
Make sure that it has gcc-9 installed. I used the following steps:

```
ck detect platform.os --platform_init_uoa=generic-linux-dummy
ck detect soft:compiler.python --full_path=`which ${PYTHON_EXE}`
ck detect soft:compiler.gcc --full_path=`which gcc-9`

ck install package --tags=mlperf,inference,src,r1.0
ck install package --tags=lib,mlperf,loadgen,static

ck install package --tags=api,model,tensorflow,r1.13.0
ck install package --tags=lib,tflite,for.coral,threads.2

ck run cmdgen:benchmark.object-detection.tflite-loadgen --verbose --library=tflite-edgetpu --model=v1 --scenario=singlestream --mode=performance --target_latency=20 --sut=rpi4coral

ck run cmdgen:benchmark.object-detection.tflite-loadgen --verbose --library=tflite-edgetpu --model=v1 --scenario=singlestream --mode=accuracy --dataset_size=5000 --sut=rpi4coral
```

Hope it helps!

**Comment by quake2005** [comment]
Thanks for your response!
It worked with me also on Ubuntu 20.04 64-bit server. However, it did work on Pi os 64-bit and 32-bit due some problems in the compiler. It has pre-installed GCC v8.3, so I tried to install GCC v9 and v10 (and G++), but It did not help.

Some other notes not related to this issue:
- I tried Ubuntu 20.04 32-bit server but it did not work as I got this massage:

```
/home/ubuntu/CK-TOOLS/mlperf-inference-r1.0/inference/loadgen/logging.h:716:26: note:   ‘const std::vector<unsigned int
’ is not derived from ‘const std::piecewise_linear_distribution<_RealType>’
  714 |     *os << ":::MLLOG {"
      |     ~~~~~~~~~~~~~~~~~~~
  715 |         << "\"key\": " << ArgValueTransform(key) << ", "
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  716 |         << "\"value\": " << ArgValueTransform(value) << ", "
      |         ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/mlperf_loadgen.dir/build.make:89: CMakeFiles/mlperf_loadgen.dir/loadgen.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/mlperf_loadgen.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   CK detected a PROBLEM in the third-party CK package:

   CK package:           lib-mlperf-loadgen-static
   CK repo:              ck-ml
   CK repo URL:          https://github.com/ctuning/ck-ml
   CK package URL:       https://github.com/ctuning/ck-ml/tree/master/package/lib-mlperf-loadgen-static
   Issues URL:           https://github.com/ctuning/ck-ml/issues

   Please, check that there is no discussion about this issue at https://cKnowledge.io/c/package/lib-mlperf-loadgen-static

   Please, submit the *full* log to:
     * https://groups.google.com/forum/#!forum/collective-knowledge
     * https://github.com/ctuning/ck/issues
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

CK error: [package] package installation failed!
```

- On CPU the instructions said that I can use `--library=tflite-v2.4.1-ruy`, however in `ck-ml/cmdgen/benchmark.object-detection.tflite-loadgen/.cm/meta.json` the last version is `tflite-v2.3-ruy` so I used it instead and it worked.

- Is there any method to run quantized model on CPU? 

Thanks!

**Comment by gfursin** [comment]
Hi!

Yes, I saw the same error on 32-bit Ubuntu so I guess it's not related to GCC ...

I managed to compile the latest MLPerf loadgen only on 64-bit machine with GCC 9+ .

I believe it's possible to fix this problem inside loadgen since v0.5 is working but I didn't have time to dig into that issue ... Maybe it will be useful to open a ticket on the MLPerf loadgen website about that. I just don't know how critical is that for MLCommons since they focus on the latest hardware and software ...

As for flite-v2.4.1-ruy - ups, I forgot to commit my update - sorry about that ;) . I just did that. If you update your ck-ml repo, you should see this change, i.e.:
```
ck pull all
```
or
```
ck pull repo:ck-ml
```
As for using quantized model, I don't know yet - there are lots of CK workflows, components and Readmes spread all over different repos including the forked ones, so I just started sorting all that, bringing them in one ck-ml repo, reproducing/testing different workflows+frameworks+models step-by-step and documenting all that to help the community. I will post notes about that whenever ready ... Please, feel free to tell us if you manage to test other models ...

Thanks!

**Comment by gfursin** [comment]
Hi @Mohamad-Jaallouk. Thank you for closing this ticket. If it's of interest, we have developed the next generation of the CK framework (CM) and are moving old CK workflows to the CM framework with the community and MLCommons within the open education workgroup: https://github.com/mlcommons/ck/blob/master/docs/mlperf-education-workgroup.md ...
